packages/core/src/request.ts

```diff
@@ -159,8 +164,50 @@ export class Request {
         this.retryCount = retryCount;
         this.errorMessages = [...errorMessages];
         this.headers = { ...headers };
-        this.userData = { ...userData };
         this.handledAt = handledAt as unknown instanceof Date ? (handledAt as Date).toISOString() : handledAt!;
+
+        Object.defineProperties(this, {
+            _userData: {
+                value: { ...userData },
+                enumerable: false,
+                writable: true,
+            },
+            userData: {
+                get: () => this._userData,
```

Comment:
I think it's a bit weird that we only define the `__crawlee` property after either the user sets `userData` or we set `skipNavigation`. Maybe I'm reading it incorrectly, but I think it works like this:

```js
const r = new Request({url: "https://example.com"});
r.userData.__crawlee // undefined
r.userData.foo = 'bar' // still undefined
r.userData = { foo: 'bar' } // now crawlee is defined
```

or

```js
const r1 = new Request({url: "https://example.com"}); // crawlee is undefined
const r2 = new Request({url: "https://example.com", skipNavigation: true }); // here it's defined
r1.skipNavigation = true; // crawlee gets defined
```

It seems kinda random to me. I was thinking that maybe we could always have at least an empty object there, but I could be missing something.

Reply:
Yes, that's how it works, its not random, its defined only once we need it. Its non enumerable and should not be used by users at all, so I don't think this really matters, but sure we can have it always defined.

If you expect we might work with it directly somewhere (or even our users), it would indeed make sense to have it always defined. But for me this is an implementation detail, not a public API.

Reply:
I had the idea of using it for other things like stats or traces, so even if we don't implement it immediately, it would make sense to have it always defined so that we can use it internally.

Reply:
Both comments handled [here](https://github.com/apify/apify-ts/commit/79f92ecf20fc8bf10c5b5fb2f3e858c8e588e376)